PasswordValidator/Validators/AdvancedValidator.cs

```diff
@@ -1,18 +1,33 @@
 ﻿using PasswordValidator.Enums;
+using PasswordValidator.Rules;
 
 namespace PasswordValidator.Validators
 {
     public class AdvancedValidator : IPasswordValidator
     {
         public ValidatorType ValidatorType => ValidatorType.Advanced;
+        private readonly IEnumerable<IRule> Rules;
+
+        public AdvancedValidator(IEnumerable<IRule> rules)
+        {
+            var lengthRule = rules.Where(rule => rule.Type == RuleType.AdvancedLength).First() ?? throw new ArgumentException($"Rule of Type {RuleType.AdvancedLength.ToString()} missing");
+            var numberRule = rules.Where(rule => rule.Type == RuleType.AdvancedNumber).First() ?? throw new ArgumentException($"Rule of Type {RuleType.AdvancedNumber.ToString()} missing");
+            var specialCharacterRule = rules.Where(rule => rule.Type == RuleType.ContainsSpecialCharacter).First() ?? throw new ArgumentException($"Rule of Type {RuleType.ContainsSpecialCharacter.ToString()} missing");
+        
+            Rules = new List<IRule> { lengthRule, specialCharacterRule, numberRule };
+        }
 
         public bool Validate(string password)
         {
-            bool isValidLength = (password.Length >= 10) && (password.Length <= 18);
-            bool containsSpecialCharacter = password.Any(character => !char.IsLetterOrDigit(character));
-            bool containsEnoughNumbers = password.Where(char.IsDigit).Count() >= 2;
+            foreach (var rule in Rules)
```

Comment:
This bit is the same for both validators - you can refactor it out into a Base Class they both share?